Description: Set __FRC_ROBORIO_ in platform config
Author: WPILib Developers <wpilib@wpi.edu>

---

--- a/gcc/config/linux.h
+++ b/gcc/config/linux.h
@@ -42,6 +42,7 @@
     do {							\
 	if (OPTION_GLIBC)					\
 	  builtin_define ("__gnu_linux__");			\
+	builtin_define ("__RASPBIAN__=10");			\
 	builtin_define_std ("linux");				\
 	builtin_define_std ("unix");				\
 	builtin_assert ("system=linux");			\
